ledger/internal/eval.go

```diff
@@ -926,13 +926,21 @@ func (eval *BlockEvaluator) Transaction(txn transactions.SignedTxn, ad transacti
 // If the transaction group cannot be added to the block without violating some constraints,
 // an error is returned and the block evaluator state is unchanged.
 func (eval *BlockEvaluator) TransactionGroup(txads []transactions.SignedTxnWithAD) error {
-	return eval.transactionGroup(txads)
+	return eval.transactionGroup(txads, nil)
+}
+
+// TransactionGroupWithTracer tentatively adds a new transaction group as part of this block evaluation,
+// calling relevant tracer hooks along the way.
+// If the transaction group cannot be added to the block without violating some constraints,
+// an error is returned and the block evaluator state is unchanged.
+func (eval *BlockEvaluator) TransactionGroupWithTracer(txads []transactions.SignedTxnWithAD, tracer logic.EvalTracer) error {
+	return eval.transactionGroup(txads, tracer)
```

Comment:
I think I was imagining putting the `tracer` in the BlockEvaluator, so the BlockEvaluator would stuff it into the eps it makes.  Could that cut down on passing into a lot of existing functions?

Reply:
That would work too, but if we do that I'd want to modify the EvalTracer functions to make tracers able to handle multiple txn groups. We have `Before/AfterInnerTxnGroup`, so we'd either need to make this capable of handling top-level groups too, or introduce a similar pair of functions.

Reply:
I think that's going to be needed very soon.

The work on Conduit is going to need to report txn by txn changes, and I'm hoping that they will get them by attaching a Tracer to the block evaluator.

Reply:
I generalized `Before/AfterInnerTxnGroup` to `Before/AfterTxnGroup` and made the tracer a field in BlockEvaluator in 0b02d9e